### PR TITLE
v3.40.0

### PR DESCRIPTION
**Pull request by murka**
https://github.com/telegraf/telegraf/discussions/1526#discussioncomment-1344548

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -1,6 +1,10 @@
 {
   "name": "telegraf",
+<<<<<<< HEAD
   "version": "3.39.0",
+=======
+  "version": "5.0.0",
+>>>>>>> develop/develop
```
Comment (wojpawlik) [inline package.json]: Resolve the conflict

File: package.json [hunk 2]
```diff
@@ -40,7 +44,7 @@
     "module-alias": "^2.2.2",
     "node-fetch": "^2.2.0",
     "sandwich-stream": "^2.0.1",
-    "typegram": "^3.1.5"
+    "typegram": "^3.4.3"
```
Comment (wojpawlik) [inline package.json]: Prefer `@grammyjs/types`

File: readme.md [hunk 3]
```diff
@@ -12,12 +12,12 @@ Bots are special [Telegram](https://telegram.org) accounts designed to handle me
...
 ### Features
 
-- Full [Telegram Bot API 4.8](https://core.telegram.org/bots/api) support
+- Full [Telegram Bot API 5.3](https://core.telegram.org/bots/api) support
 - [Telegram Payment Platform](https://telegram.org/blog/payments)
 - [HTML5 Games](https://core.telegram.org/bots/api#games)
 - [Inline mode](https://core.telegram.org/bots/api#inline-mode)
 - Incredibly fast
-- [now](https://now.sh)/[Firebase](https://firebase.google.com/products/functions/)/[Glitch](https://dashing-light.glitch.me)/[Heroku](https://devcenter.heroku.com/articles/getting-started-with-nodejs#introduction)/[AWS **λ**](https://docs.aws.amazon.com/lambda/latest/dg/nodejs-prog-model-handler.html)/Whatever ready
+- [Vercel](https://vercel.com)/[Firebase](https://firebase.google.com/products/functions/)/[Glitch](https://dashing-light.glitch.me)/[Heroku](https://devcenter.heroku.com/articles/getting-started-with-nodejs#introduction)/[AWS **λ**](https://docs.aws.amazon.com/lambda/latest/dg/nodejs-prog-model-handler.html)/Whatever ready
```
Comment (wojpawlik) [inline readme.md:20]: I don't think it's possible to deploy on Vercel now
```suggestion
- [Firebase](https://firebase.google.com/products/functions/)/[Glitch](https://dashing-light.glitch.me)/[Heroku](https://devcenter.heroku.com/articles/getting-started-with-nodejs#introduction)/[AWS **λ**](https://docs.aws.amazon.com/lambda/latest/dg/nodejs-prog-model-handler.html)/Whatever ready
```

File: readme.md [hunk 4]
```diff
@@ -1,5 +1,5 @@
 ![Telegraf](docs/media/header.png)
```
Comment (wojpawlik) [inline readme.md:1]: I made svg logo
```suggestion
<img src="https://raw.githubusercontent.com/telegraf/telegraf/v4/docs/assets/logo.svg" alt="logo" height="90" align="left">
<h1 style="display: inline">telegraf.js</h1>

Modern Telegram Bot API framework for Node.js
```

File: typings/context.d.ts [hunk 5]
```diff
@@ -103,11 +103,17 @@ export declare class TelegrafContext {
   getChatMember(userId: number): Promise<tt.ChatMember>
 
   /**
-   * Use this method to get the number of members in a chat
+   * Legasy, see getChatMemberCount
```
Comment (wojpawlik) [inline typings/context.d.ts:106]:
```suggestion
   * @deprecated in favor of `getChatMemberCount`
```

File: typings/telegram-types.d.ts [hunk 6]
```diff
@@ -661,10 +672,45 @@ export type Extra = ExtraSendMessage
...
+   * Applied for supergroups and channels only.
+   */
+   until_date?: number,
+
+   /**
+    * Pass True to delete all messages from the chat for the user that is being removed.
+    * If False, the user will be able to see messages in the group that were sent before the user was removed.
+    * Always True for supergroups and channels.
+    */
+    revoke_messages: boolean
```
Comment (wojpawlik) [inline typings/telegram-types.d.ts:694]: Please add prettier to ensure consistent indentation

File: typings/telegram.d.ts [hunk 7]
```diff
@@ -467,12 +467,19 @@ export declare class Telegram extends ApiClient {
   getChatMember(chatId: string | number, userId: number): Promise<tt.ChatMember>
 
   /**
-   * Use this method to get the number of members in a chat
+   * Legasy, see getChatMemberCount
```
Comment (wojpawlik) [inline typings/telegram.d.ts:470]:
```suggestion
   * @deprecated in favor of `getChatMemberCount`
```

File: typings/telegram.d.ts [hunk 8]
```diff
@@ -499,7 +506,20 @@ export declare class Telegram extends ApiClient {
...
+   * @returns True on success
+   */
+   banChatMember(
+    chatId: number | string,
+    userId: number,
+    extra?: tt.ExtraBan
+  ): Promise<boolean>
+
+  /**
+   * Legacy, see banChatMember
```
Comment (wojpawlik) [inline typings/telegram.d.ts:522]:
```suggestion
   * @deprecated in favor of `banChatMember`
```

**Review by wojpawlik — CHANGES_REQUESTED** [review]
I don't see any significant breaking changes, I'd rather this was 3.40, and Telegraf 5 addressed some [breaking issues](https://github.com/telegraf/telegraf/issues?q=is%3Aissue+is%3Aopen+label%3Abreaking).

**Comment by wojpawlik** [comment]
Please add a [CI workflow](https://github.com/telegraf/telegraf/blob/v4/.github/workflows/node.js.yml)

File: readme.md (hunk 3 above)
Comment (murka) [reply readme.md:20]: Why not? I have bots on Vercel right now, I use [Serverless Functions](https://vercel.com/docs/functions/introduction)

Comment (wojpawlik) [reply readme.md:20]: Then please link to a guide instead. It was not straightforward anymore last time I checked.

Comment (murka) [reply readme.md:20]: I have not used anything other than [Serverless Function](https://vercel.com/docs/functions/introduction) and [Telegram Bot API](https://core.telegram.org/bots/api). I think, I can write a guide for you as soon as possible.

**Comment by wojpawlik** [comment]
For v5, I'd set `engines.node` (minimum Node version) to `>=14.13.1` (optional chaining, nullish coalescing) or `>=16.7.0` (`AbortController`, `Blob::stream()`).

File: package.json (hunk 1 above)
Comment (murka) [reply package.json]: I think, you can move to contributing for grammyjs. We deprecate v4, and publish this version as v5 after fix all bugs on this version.

Comment (wojpawlik) [reply package.json]: Why not release this as v3.40.0, and address design flaws afterwards, in v5.0.0?

@MKRhere might be interested in maintaining v4.

Comment (MKRhere) [reply package.json]: Indeed

**Comment by MKRhere** [comment]
Good to see work on this @murka; since this is a non-breaking branch and the idea is to keep v3 updated for those who prefer it over v4, why not target v3.40.0, as I'm working on fixing v4 issues in a new breaking release as v5?

**Comment by murka** [comment]
@MKRhere, you are welcome, so I don't know what version will be approved, v5 is a draft name

**Comment by MKRhere** [comment]
Whenever you're ready to merge, please mark this PR as ready to review

**Pull request closed without merging**